src/pip/_internal/cache.py

```diff
@@ -58,6 +58,10 @@ def _get_cache_path_parts(self, link):
         key_parts = [link.url_without_fragment]
         if link.hash_name is not None and link.hash is not None:
             key_parts.append("=".join([link.hash_name, link.hash]))
+        if link.subdirectory_fragment:
+            key_parts.append(
+                "=".join(["subdirectory", link.subdirectory_fragment])
+            )
         key_url = "#".join(key_parts)
```

Comment:
While this is a generic key so technically it doesn't have to be a valid URL, I think it would probably be nice to join the fragment parts with `&` then join those to the first part with `#`. It would probably be easier to relay if we actually had to communicate this info to anyone. :)

Reply:
I definitely like this idea which reminds me of https://github.com/pypa/pip/pull/3734/files#diff-2695f32c4432acd141c3dbe7e7e3a6b0R884-R888 which was exposed via my late `pip cache` command :thinking: .

Reply:
Note in #7319 the key we are hashing becomes even less URLish.
So could we just concatenate key parts with an arbitrary separator before hashing?

Reply:
Good point, I'll make a comment over there. Given that #7319 will invalidate existing caches, we could probably refactor prior to that change to change the cache key construction to something more regular. I would just leave this as-is for this PR (joining with "#") and we can address the general issue separately.